ShareX.HelpersLib/Helpers/ImageHelpers.cs

```diff
@@ -1768,53 +1768,57 @@ public static Bitmap TornEdges(Bitmap bmp, int tornDepth, int tornRange, AnchorS
 
             if (sides.HasFlag(AnchorStyles.Top) && horizontalTornCount > 1)
             {
-                for (int x = 0; x < horizontalTornCount - 1; x++)
+                for (int x = 0; x < bmp.Width; x += tornRange)
                 {
-                    points.Add(new Point(tornRange * x, RandomFast.Next(0, tornDepth)));
+                    int y = random ? RandomFast.Next(0, tornDepth) : ((x / tornRange) & 1) * tornDepth;
+                    points.Add(new Point(x, y));
                 }
             }
             else
             {
                 points.Add(new Point(0, 0));
-                points.Add(new Point(bmp.Width - 1, 0));
+                points.Add(new Point(bmp.Width, 0));
```

Comment:
Wouldn't `bmp.Width` be outside of bitmap? Because index is 0 based.

Reply:
It doesn't produce errors. Also I'm pretty sure the paths are left-inclusive right-exclusive (and similar top-inclusive bottom-exclusive).

Reply:
You can draw outside, it won't throw error.

> left-inclusive right-exclusive

What is that even means.

Reply:
I tried with and without -1 on the far edges. The result is definitely more wrong with it than without it, but it's wrong even without it too.
For the first two of these images, I cut out the top edge of the image with Wave Edge effect. The third image was opened in the image editor and then saved with no changes made. The first has -1 on Width and Height coordinates for the effect-less edges, the second uses Width and Height as given.
After saving the images from ShareX, I opened them in Krita, added a fill layer underneath, and saved again.

It seems that the pixel coordinate model is in fact center of pixel, so (0, 0) is the center of the top left pixel, meaning that a rectangle starting at (0, 0) excludes half of the top row and half of the left row of pixels. For correct, seamless handling (with float-coordinates only?) the top left of the top left pixel then has to be addressed as (-0.5, -0.5), and the right edge of the rightmost pixel has coordinate (Width - 0.5).

<img width="599" alt="devenv_2022-08-17_23-52-46 bg" src="https://user-images.githubusercontent.com/1062071/185250965-f406154a-4abc-4ee0-9654-914c011ce030.png">
<img width="592" alt="devenv_2022-08-17_23-53-42 bg" src="https://user-images.githubusercontent.com/1062071/185250967-04afa81c-62d1-46d3-8173-06a94256cc35.png">
<img width="578" alt="devenv_2022-08-17_23-57-40 bg" src="https://user-images.githubusercontent.com/1062071/185250970-42bb4120-26ac-472c-81c4-966b805524a7.png">

Reply:
Yep, offsetting everything by -0.5 makes it hit the image edges perfectly.

![image](https://user-images.githubusercontent.com/1062071/185344977-d899bac3-776b-4dc3-ad59-d34464189fe8.png)

Reply:
https://github.com/ShareX/ShareX/search?q=PixelOffsetMode

Reply:
That certainly makes things easier.

The coordinates still need to use bmp.Width and bmp.Height to be correct and not leave a one-pixel edge on the right and bottom.

(On a side note, I find it weird that Microsoft never directly documents what the pixel coordinate system for antialiasing is, or at least I can't find it mentioned. Oh well.)